[Do not Merge in 0.14] Implementation Dot Plots

JTPetter opened the pull request:
Fixes https://github.com/jasp-stats/jasp-issues/issues/549
Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1056

JTPetter committed:
Implementation Dot Plots

Fixes https://github.com/jasp-stats/INTERNAL-jasp/issues/1056

vandenman commented on JASP-Engine/JASP/R/descriptives.R:
```diff
@@ -1055,6 +1078,56 @@ Descriptives <- function(jaspResults, dataset, options) {
+    split <- names(dataset)
+    
+    plotResult <- createJaspContainer(title=variable)
+    plotResult$dependOn(options="splitby", optionContainsValue=list(variables=variable))
```
```suggestion
    plotResult$dependOn(optionContainsValue=list(variables=variable))
```
since we have `jaspResults[["DotPlots"]]$dependOn(c("splitby", "descriptivesDotPlot"))`, this already depends on `"splitby"`.

vandenman commented on JASP-Engine/JASP/R/descriptives.R:
```diff
@@ -195,6 +195,29 @@ Descriptives <- function(jaspResults, dataset, options) {
+    
+    if(is.null(jaspResults[["DotPlots"]])) {
+      jaspResults[["DotPlots"]] <- createJaspContainer(gettext("Dot Plots"))
+      jaspResults[["DotPlots"]]$dependOn(c("plotVariables", "splitby", "descriptivesDotPlot"))
```
```suggestion
      jaspResults[["DotPlots"]]$dependOn(c("splitby", "descriptivesDotPlot"))
```
`plotVariables` is not in the options right? If so, depending on it has no effect so I'd just omit it.

vandenman commented on JASP-Engine/JASP/R/descriptives.R:
```diff
@@ -1055,6 +1078,56 @@ Descriptives <- function(jaspResults, dataset, options) {
+    plotResult <- createJaspContainer(title=variable)
+    plotResult$dependOn(options="splitby", optionContainsValue=list(variables=variable))
+    
+    for (l in split) {
+      plotResult[[l]] <- .descriptivesDotPlots_SubFunc(dataset=dataset[[l]], variable=variable, title = l)
+    }
```
```suggestion
    for (l in split)
      plotResult[[l]] <- .descriptivesDotPlots_SubFunc(dataset = dataset[[l]], variable = variable, title = l)
```
Two nitpicks:
1. Braces `{` arent necessary for a single line for loop
2. Please be consistent with spaces around arguments (e.g., `dataset = dataset[[l]]`). I know that doesn't happen everywhere in JASP, but that's a bad thing.

vandenman commented on JASP-Engine/JASP/R/descriptives.R:
```diff
@@ -1055,6 +1078,56 @@ Descriptives <- function(jaspResults, dataset, options) {
+  
+  x <- na.omit(dataset[[.v(variable)]])
+  
+  yheight <- max(as.data.frame(table(x))$Freq)
```
1. why the `as.data.frame`?
2. `yheight` is unused?

vandenman commented on JASP-Engine/JASP/R/descriptives.R:
```diff
@@ -195,6 +195,29 @@ Descriptives <- function(jaspResults, dataset, options) {
+    
+    for (var in variables) {
+      if(is.null(dotPlots[[var]])) {
+        if (makeSplit) {
+          dotPlots[[var]] <- .descriptivesDotPlots(dataset = splitDat.factors, options = options, variable = var)
+        } else {
+          dotPlots[[var]] <- .descriptivesDotPlots(dataset = dataset.factors, options = options, variable = var)
+          
+        }
```
so this is perfectly fine, but you could also write:
```r
dotPlots[[var]] <- .descriptivesDotPlots(
  dataset  = if (makeSplit) splitDat.factors else dataset.factors, 
  options  = options, 
  variable = var
)
```
That way it is ever so slightly more clear that the if statement only affects the first argument.

vandenman reviewed (changes requested):
The functionality looks great! I found one corner case:

![image](https://user-images.githubusercontent.com/21319932/95482976-fbd55400-098e-11eb-849a-7cb28378fab9.png)

Ideally, we'd show all possible levels on the x-axis here (even if the level doesn't appear for a particular split). It also looks like something goes wrong for some plots with only one level. For example, given `debString` = b there are 3 occurences of `f` and 0 of `m` (tick Frequency tables to see this info). But the plot shows 0 dots.

Aside from that, I left some comments about the code below.

JTPetter committed:
Fix plotting empty categories on x-axis

-Fix to plot all categories on x-axis also when the catgeory is empty
-Style changes

vandenman commented on JASP-Engine/JASP/R/descriptives.R:
```diff
@@ -1055,6 +1076,65 @@ Descriptives <- function(jaspResults, dataset, options) {
+  }else{
+    
+    dotplot <- .descriptivesDotPlots_SubFunc(dataset = dataset, variable = variable, title = variable)
+    dotplot$dependOn(options="splitby", optionContainsValue = list(variables = variable))
```
```suggestion
    dotplot$dependOn(optionContainsValue = list(variables = variable))
```
Since we have `jaspResults[["DotPlots"]]$dependOn(c("splitby", "descriptivesDotPlot"))`, the dotplot already depends on `"splitby"`, so this dependency is superfluous.

vandenman reviewed (changes requested):
Code & Functionality looks great! One final comment, see below.

JTPetter committed:
Redundant dependency

Co-authored-by: Don van den Bergh <donvdbergh@hotmail.com>

TimKDJ commented:
This PR is continued here: https://github.com/jasp-stats/jaspDescriptives/pull/5
